codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -213,33 +213,53 @@ class KotlinJmespathExpressionVisitor(
         return acceptSubexpression(this.arguments[0]) to acceptSubexpression(this.arguments[1])
     }
 
+    private fun VisitedExpression.dotFunction(expression: FunctionExpression, expr: String, elvisExpr: String? = null): VisitedExpression {
+        val dotFunctionExpr = ensureNullGuard(this.shape, expr, elvisExpr)
+        val ident = addTempVar(expression.name, "${this.identifier}$dotFunctionExpr")
+
+        return VisitedExpression(ident, this.shape)
+    }
+
     override fun visitFunction(expression: FunctionExpression): VisitedExpression = when (expression.name) {
         "contains" -> {
             val (subject, search) = expression.twoArgs()
-
-            val containsExpr = ensureNullGuard(subject.shape, "contains(${search.identifier})", "false")
-            val ident = addTempVar("contains", "${subject.identifier}$containsExpr")
-
-            VisitedExpression(ident)
+            subject.dotFunction(expression, "contains(${search.identifier})", "false")
         }
 
         "length" -> {
             writer.addImport(RuntimeTypes.Core.Utils.length)
             val subject = expression.singleArg()
-
-            val lengthExpr = ensureNullGuard(subject.shape, "length", "0")
-            val ident = addTempVar("length", "${subject.identifier}$lengthExpr")
-
-            VisitedExpression(ident)
+            subject.dotFunction(expression, "length", "0")
         }
 
         "abs", "floor", "ceil" -> {
             val number = expression.singleArg()
+            number.dotFunction(expression, "let { kotlin.math.${expression.name}(it.toDouble()) }")
+        }
 
-            val functionExpr = ensureNullGuard(number.shape, "let { kotlin.math.${expression.name}(it.toDouble()) }")
-            val ident = addTempVar(expression.name, "${number.identifier}$functionExpr")
+        "sum" -> {
+            val numbers = expression.singleArg()
+            numbers.dotFunction(expression, "sum()")
+        }
+
+        "avg" -> {
+            val numbers = expression.singleArg()
+            numbers.dotFunction(expression, "average()")
+        }
```

Comment:
**Correctness**: I don't think this matches the [JMESPath spec for `avg`](https://jmespath.org/specification.html#avg) given an empty array. It should return `null` but I believe this returns `NaN`.